lld/ELF/Driver.cpp

```diff
@@ -2619,6 +2623,42 @@ static uint32_t getAndFeatures() {
   return ret;
 }
 
+static void getAArch64PauthInfo() {
+  auto it = std::find_if(
+      ctx.objectFiles.begin(), ctx.objectFiles.end(),
+      [](const ELFFileBase *f) { return !f->aarch64PauthAbiTag.empty(); });
+  if (it == ctx.objectFiles.end())
+    return;
+
+  ctx.aarch64PauthAbiTag = (*it)->aarch64PauthAbiTag;
+  StringRef f1 = (*it)->getName();
+  for (ELFFileBase *f : ctx.objectFiles) {
+    StringRef f2 = f->getName();
+    ArrayRef<uint8_t> d1 = ctx.aarch64PauthAbiTag;
+    ArrayRef<uint8_t> d2 = f->aarch64PauthAbiTag;
+    if (d2.empty()) {
+      auto helper = [](StringRef report, const Twine &msg) {
+        if (report == "warning")
+          warn(msg);
+        else if (report == "error")
+          error(msg);
+      };
+
+      helper(config->zPauthReport,
```

Comment:
Per `checkAndReportMissingFeature`, the message can be 
`-z pauth-report: file does not have ...`

`either all or no input files must have it` is probably unnecessary.

Reply:
I've moved the `if` statement body from `checkAndReportMissingFeature` to `reportMissingFeature` and used that here instead of `auto helper`. Using just `checkAndReportMissingFeature` is not applicable here since it's designed for `andFeatures` stored as bitmask - the "check" part is `if (f->aarch64PauthAbiTag.empty())` in our case.

See 1448a4a4e0e3cb70358558579e97fce3f0990356